### PR TITLE
Support pagination for access with cache

### PR DESCRIPTION
**Pull request by astrozzc**
## Link(s) to Jira
- https://issues.redhat.com/browse/RHCLOUD-7455

## Description of Intent of Change(s)
The limit and offset are ignored after cache is enabled. This
is to support pagination again by adding limit and offset into
the sub key.
Then the record will be saved into each entry "principal" and sub
key, e.g. "catalog_0_1000"

## Local Testing
Unit tests
Put some roles into the db, setup server locally and run each of the following twice:
curl -v GET "http://0.0.0.0:8080/r/insights/platform/rbac/v1/access/?application=&limit=20" -w '\nTotal: %{time_total}s\n'
curl -v GET "http://0.0.0.0:8080/r/insights/platform/rbac/v1/access/?application=" -w '\nTotal: %{time_total}s\n'
curl -v GET "http://0.0.0.0:8080/r/insights/platform/rbac/v1/access/?application=catalog&offset=4" -w '\nTotal: %{time_total}s\n'

You will see the total time is different for each run to make sure the caching is enabled.

## Checklist
- [ ] if API spec changes are required, is the spec updated?
- [ ] are there any pre/post merge actions required? if so, document here.
- [x] are theses changes covered by unit tests?
- [ ] if warranted, are documentation changes accounted for?
- [ ] does this require migration changes?
  - [ ] if yes, are they backwards compatible?
- [ ] is there known, direct impact to dependent teams/components?
  -  if yes, how will this be handled?

## Secure Coding Practices Checklist Link
- https://github.com/RedHatInsights/secure-coding-checklist

## Secure Coding Practices Checklist
- [ ] Input Validation
- [ ] Output Encoding
- [ ] Authentication and Password Management
- [ ] Session Management
- [ ] Access Control
- [ ] Cryptographic Practices
- [ ] Error Handling and Logging
- [ ] Data Protection
- [ ] Communication Security
- [ ] System Configuration
- [ ] Database Security
- [ ] File Management
- [ ] Memory Management
- [ ] General Coding Practices

### CODE REVIEW

File: rbac/management/access/view.py [hunk 1]
```diff
@@ -117,3 +118,15 @@ def get_paginated_response(self, data):
         """Return a paginated style `Response` object for the given output data."""
         assert self.paginator is not None
         return self.paginator.get_paginated_response(data)
+
+    def generate_sub_key(self, request):
+        """Generate the sub key to store/retrive record from redis."""
```
Comment (catastrophe-brandon) [inline rbac/management/access/view.py:123]: minor typo: retrive -> retrieve

Comment (astrozzc) [reply rbac/management/access/view.py]: Good catch, thanks 😁

**Comment by astrozzc** [comment]
Oops, I thought the test failure is related to my test but I just realize that it is because there is no redis cache for our travis and pr check. Will discuss with our engineers to find a solution.

File: rbac/management/access/view.py [hunk 2]
```diff
@@ -117,3 +118,17 @@ def get_paginated_response(self, data):
...
+
+    def generate_sub_key(self, request):
+        """Generate the sub key to store/retrieve record from redis."""
+        query_params = request.query_params
+        app = query_params.get(APPLICATION_KEY)
+        if not app:
+            app = "all"
+        limit = int(query_params.get("limit", self.paginator.default_limit))
+        # If there are some team setting this out of range, set it to the max support
+        if limit > self.paginator.max_limit:
```
Comment (coderbydesign) [inline rbac/management/access/view.py:130]: Good call!

**Review by coderbydesign — APPROVED** [review]
Looks good to me!

**Review by wcmitchell — APPROVED** [review]
LGTM!

**Pull request merged**